handlers/rsa.go

```diff
@@ -0,0 +1,211 @@
+		return true, "please, do not share the key"
+	}
+	if rh.exist(voterId) {
+		log.Warnf("%s already registered", voterId)
+		return false, "already registered"
+	}
+
+	rh.addKey(voterId, nil)
+	log.Infof("new user registered with id %s", voterId)
+
+	return true, ""
+}
+
+// RequireCertificate must return true if the auth handler requires some kind of client
+// TLS certificate. If true then CertificateCheck() and HardcodedCertificate() methods
+// must be correctly implemented. Else both function can just return true and nil.
+func (rh *RsaHandler) RequireCertificate() bool {
+	return false
+}
+
+// CertificateCheck is used by the Auth handler to ensure a specific certificate is
+// added to the CA cert pool on the HTTP/TLS layer (optional).
+func (rh *RsaHandler) CertificateCheck(subject []byte) bool {
+	return true
+}
+
+// Certificates returns a hardcoded CA certificated that will be added to the
+// CA cert pool by the handler (optional).
+func (rh *RsaHandler) Certificates() [][]byte {
+	return nil
+}
+
+// Internal data handlers
+
+func parseRsaPublicKey() (*rsa.PublicKey, error) {
+	if len(rsaPubKeys) < 1 {
+		return nil, errors.Errorf("no public keys")
+	}
+	// Using the first one available so far
+	block, _ := pem.Decode([]byte(rsaPubKeys[0]))
+	if block == nil {
+		log.Warnf("failed to parse the public key")
+		return nil, errors.Errorf("failed to parse the public key")
+	}
+
+	parsedKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	switch parsedKey.(type) {
+	case *rsa.PublicKey:
+		break
+	case *dsa.PublicKey:
+	case *ecdsa.PublicKey:
+	case ed25519.PublicKey:
+	default:
+		log.Warnf("cannot parse the public key")
+		return nil, errors.Errorf("cannot parse the public key")
+	}
+	return parsedKey.(*rsa.PublicKey), nil
+}
+
+// parseRsaAuthData transforms the incoming authData string array and returns a digested output
+// of the relevant parameters for the handler
+func parseRsaAuthData(authData []string) ([]byte, []byte, []byte, error) {
+	if len(authData) != 3 {
+		log.Warnf("invalid params (3 items expected)")
+		return nil, nil, nil, errors.Errorf("invalid params (3 items expected)")
+	}
+
+	// Catenate hex
+	processId := authData[0]
+	if len(processId) != 64 {
+		log.Warnf("invalid electionId")
+		return nil, nil, nil, errors.Errorf("invalid electionId")
+	}
+	voterId := authData[1]
+	if len(voterId) != 64 {
+		log.Warnf("invalid voterId")
+		return nil, nil, nil, errors.Errorf("invalid voterId")
+	}
+
+	voterIdBytes, err := hex.DecodeString(voterId)
+	if err != nil || len(voterIdBytes) != 32 {
+		log.Warnf("invalid voterId: %s", voterId)
+		return nil, nil, nil, errors.Errorf("invalid voterId")
+	}
+
+	message, err := hex.DecodeString(processId + voterId)
+	if err != nil || len(message) != 32+32 {
+		// By discard, only processId can be invalid
+		log.Warnf("invalid electionId: %s", processId)
+		return nil, nil, nil, errors.Errorf("invalid electionId")
```

Comment:
When returning an error that comes from another error, its a good practice to wrap it: `fmt.Errorf("invalid electionId: %w", err)`

Reply:
Done
Wouldn't the extra `log.Warnf` be redundant here?

Reply:
Yes